metagraph/src/graph/alignment/aligner_seeder_methods.cpp

```diff
@@ -81,57 +84,234 @@ void ExactSeeder<NodeType>::call_seeds(std::function<void(Seed&&)> callback) con
+            // the reverse complement of the sub-k match will fall somewhere in this range
+            size_t j_min = query_rc.size() - i - max_seed_length;
+            size_t j_max = query_rc.size() - i - this->config_.min_seed_length;
+
+            // skip over positions which have better matches
+            while (min_seed_length[j_min] > max_seed_length && j_min <= j_max) {
+                ++j_min;
+                --max_seed_length;
             }
+
+            if (j_min > j_max)
+                continue;
+
+            const auto &boss = dbg_succ_.get_boss();
+
+            auto encoded = boss.encode({ query_rc.data() + i, max_seed_length });
+            auto [first, last, end] = boss.index_range(encoded.begin(), encoded.end());
+
+            size_t seed_length = end - encoded.begin();
+            size_t j = query_rc.size() - i - seed_length;
+
+            assert(seed_length < this->config_.min_seed_length
+                || j < min_seed_length.size());
+
+            if (seed_length < this->config_.min_seed_length
+                    || seed_length < min_seed_length[j]) {
+                continue;
+            }
+
+            min_seed_length[j] = seed_length;
+
+            // clear out shorter seeds
+            size_t s = seed_length;
+            for (size_t m = j + 1;
+                    m < min_seed_length.size() && s > min_seed_length[m];
+                    ++m) {
+                assert(suffix_seeds[m].empty()
+                    || min_seed_length[m] == this->graph_.get_k()
+                                                - suffix_seeds[m][0].get_offset());
+                min_seed_length[m] = s--;
+                suffix_seeds[m].clear();
+            }
+
+            // e.g., match: $$$ATG, want ATG$$$
+            suffix_to_prefix(dbg_succ_, std::make_tuple(first, last, seed_length),
+                             [&](node_index prefix_node) {
+                append_suffix_seed(
+                    j, canonical->reverse_complement(prefix_node), seed_length
+                );
+            });
+        }
+    }
+
+    // once all seeds are aggregated, run callback
+    for (size_t i = 0; i < suffix_seeds.size(); ++i) {
+        std::vector<Seed> &pos_seeds = suffix_seeds[i];
+        if (pos_seeds.empty())
+            continue;
+
+        if (!pos_seeds[0].get_offset()) {
+            assert(min_seed_length[i] == this->graph_.get_k());
+            assert(pos_seeds.size() == 1);
+            callback(std::move(pos_seeds[0]));
         } else {
-            last_seed_size = std::max(this->graph_.get_k() - 1,
-                                      this->config_.min_seed_length);
+            assert(min_seed_length[i] == this->graph_.get_k() - pos_seeds[0].get_offset());
+            if (pos_seeds.size() <= this->config_.max_num_seeds_per_locus) {
+                for (auto&& seed : pos_seeds) {
+                    callback(std::move(seed));
+                }
+            }
         }
     }
```

Comment:
if you anyway collect them into a vector, just return the vector.
There is no point in using a callback.